# Shopify - Winter 2021 Data Science Intern Challenge - Matthew Law
## Question 1
### Part (a)

In [1]:
import pandas as pd

In [2]:
dataset=pd.read_csv('2019 Winter Data Science Intern Challenge Data Set - Sheet1.csv')

In [3]:
dataset.head()

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11

In [4]:
round(dataset['order_amount'].mean(),2)

3145.13

The arithmetic mean is taken from the [order_amount] column to produce the AOV of $3145.13. 

In [5]:
dataset['order_amount'].skew()

16.675033432503792

Although the [order_amount] data is continuous, the disadvantage in using the mean as a measure of central tendency is that it is susceptible to skewing by outliers in the data. The skew() function produced a largely positive number, this means that the order_amount

In [9]:
dataset.loc[dataset['order_amount']>=3145.13].shape[0]

63

In [10]:
(dataset.loc[dataset['order_amount']>=3145.13].shape[0]/dataset['order_amount'].shape[0])*100

1.26

We can see that 63/5000 (1.26%) of the orders are above or equal to the original AOV of $3145.13. This tells us that the arithmetic mean is likely not an appropriate measure of central tendency, which is meant to describe a set of data with a central position.

In [11]:
user_analysis = dataset.groupby(['user_id']).mean()[['total_items', 'order_amount']]

user_analysis.head(10)

total_items   order_amount
user_id                            
607      2000.000000  704000.000000
700         1.875000     299.375000
701         2.615385     397.076923
702         2.615385     406.615385
703         2.562500     380.687500
704         1.785714     275.285714
705         2.000000     306.000000
706         1.846154     252.230769
707         1.666667    3087.222222
708         2.071429     321.285714

In [12]:
shop_analysis = dataset.groupby(['shop_id']).sum()[['order_amount', 'total_items']]

shop_analysis['Avg_Cost_Per_Item'] = shop_analysis['order_amount']/shop_analysis['total_items']

shop_analysis[['Avg_Cost_Per_Item']].sort_values(by='Avg_Cost_Per_Item', ascending=False).head(10)

Avg_Cost_Per_Item
shop_id                   
78                 25725.0
42                   352.0
12                   201.0
89                   196.0
99                   195.0
50                   193.0
38                   190.0
51                   187.0
6                    187.0
11                   184.0

Although the [order_amount] data is continuous, the disadvantage in using the mean as a measure of central tendency is that it is susceptible to skewing by outliers in the data. Through another spot check of the raw data, we can determine that the outliers in the [order_amount] column are a product of: 

    -A user (user_id = 607) who placed 17 separate orders of 2000 total_items totaling 704000 from a shop (shop_id = 42)

    -A shop (shop_id = 78) that sells an item priced at $25725 per unit.

This results in a right skewed dataset. Taking into consideration the product type (shoes), a usually relatively affordable item, it must be decided whether bulk buyers and shoes that are priced well above the average should be included in the AOV metric, assuming that this data was not recorded in error.

### Part (b)

The most appropriate measure of central tendency for this dataset would be median. Through part (a), we see that the data is right skewed. This means that the frequency distribution for the data is right skewed, and the mean will shift towards the right, in the direction of the skew. The median is not as easily influenced by skewed data, and will prevent bulk orders and outlier prices from giving a false representation of a typical shoe order.

In comparison to the other measures of central tendency:
-Mean would not be appropriate because of the aforementioned reasons in part (a)
-Mode would be appropriate for categorical data, but [order_amount] data is continuous.


### Part (c)

In [19]:
round(dataset['order_amount'].median(),2)

284.0

The value of this new suggested metric of median order value (MOV) is $284, which is a better representation of a typical shoe order.

## Question 2
### Part (a) How many orders were shipped by Speedy Express in total?

SELECT COUNT(DISTINCT [OrderID]) AS [Num_Orders_Shipped_By_Speedy_Express] \
&emsp;FROM [Orders] O JOIN [Shippers] S ON O.[ShipperID] = S.[ShipperID] \
&emsp; &emsp;AND [ShipperName] = 'Speedy Express';


'54' orders were shipped by Speedy Express in total.

### Part (b) What is the last name of the employee with the most orders?

SELECT E.[LastName] AS [Employee_Last_Name_Most_Orders]  \
FROM [OrderDetails] OD \
&emsp;JOIN [Orders] O ON OD.[OrderID] = O.[OrderID] \
&emsp;JOIN [Employees] E ON O.[EmployeeID] = E.[EmployeeID] \
GROUP BY O.[EmployeeID] \
ORDER BY COUNT(*) DESC \
LIMIT 1;


'Peacock' is the name of the employee with the most orders.

### Part (b) What product was ordered the most by customers in Germany?

SELECT [ProductName] AS [Most_Popular_Product_Germany] \
FROM [OrderDetails] OD \
&emsp;JOIN [Products] P ON OD.[ProductID] = P.[ProductID] \
&emsp;JOIN [Orders] O ON OD.[OrderID] = O.[OrderID] \
&emsp;JOIN [Customers] C ON O.[CustomerID] = C.[CustomerID] \
WHERE Country = 'Germany' \
GROUP BY P.[ProductID] \
ORDER BY SUM([Quantity]) DESC \
LIMIT 1;

'Boston Crab Meat' was the product most ordered by customers in Germany.